key.core/src/main/java/de/uka/ilkd/key/logic/op/LocationVariable.java

```diff
@@ -86,4 +86,29 @@ public int hashCodeModProofIrrelevancy() {
             argSorts(), name().toString(), arity(),
             whereToBind(), isRigid());
     }
+
+    @Override
+    public boolean equals(Object obj) {
```

Comment:
I think that KeY has been such that references to a program variable only refer to the same variable if they use the same location variable operator object.

If that is really to be changed, I really think the entire code base must be inspected to change this fundamental decision.